sklearn/cluster/_hdbscan/_linkage.pyx

```diff
@@ -1,210 +1,259 @@
+
+        MST_edge_dtype = np.dtype([
+            ("current_node", np.int64),
+            ("next_node", np.int64),
+            ("distance", np.float64),
+        ])
+    """
     cdef:
-        cnp.ndarray[cnp.intp_t, ndim=1] node_labels
-        cnp.ndarray[cnp.intp_t, ndim=1] current_labels
-        cnp.ndarray[cnp.double_t, ndim=1] current_distances
-        cnp.ndarray[cnp.double_t, ndim=1] left
-        cnp.ndarray[cnp.double_t, ndim=1] right
-        cnp.ndarray[cnp.double_t, ndim=2] result
-
-        cnp.ndarray label_filter
-
-        cnp.intp_t current_node
-        cnp.intp_t new_node_index
-        cnp.intp_t new_node
-        cnp.intp_t i
-
-    result = np.zeros((distance_matrix.shape[0] - 1, 3))
-    node_labels = np.arange(distance_matrix.shape[0], dtype=np.intp)
+        cnp.ndarray[cnp.int64_t, ndim=1, mode='c'] current_labels
+        cnp.ndarray[cnp.float64_t, ndim=1, mode='c'] min_reachability, left, right
+        cnp.ndarray[MST_edge_t, ndim=1, mode='c'] mst
+
+        cnp.ndarray[cnp.uint8_t, mode='c'] label_filter
+
+        cnp.int64_t n_samples = mutual_reachability.shape[0]
+        cnp.int64_t current_node, new_node_index, new_node, i
+
+    mst = np.empty(n_samples - 1, dtype=MST_edge_dtype)
+    current_labels = np.arange(n_samples, dtype=np.int64)
     current_node = 0
-    current_distances = np.infty * np.ones(distance_matrix.shape[0])
-    current_labels = node_labels
-    for i in range(1, node_labels.shape[0]):
+    min_reachability = np.infty * np.ones(n_samples, dtype=np.float64)
+    for i in range(0, n_samples - 1):
         label_filter = current_labels != current_node
         current_labels = current_labels[label_filter]
-        left = current_distances[label_filter]
-        right = distance_matrix[current_node][current_labels]
-        current_distances = np.where(left < right, left, right)
+        left = min_reachability[label_filter]
+        right = mutual_reachability[current_node][current_labels]
+        min_reachability = np.minimum(left, right)
 
-        new_node_index = np.argmin(current_distances)
+        new_node_index = np.argmin(min_reachability)
         new_node = current_labels[new_node_index]
-        result[i - 1, 0] = <double> current_node
-        result[i - 1, 1] = <double> new_node
-        result[i - 1, 2] = current_distances[new_node_index]
+        mst[i].current_node = current_node
+        mst[i].next_node = new_node
+        mst[i].distance = min_reachability[new_node_index]
         current_node = new_node
 
-    return result
+    return mst
 
 
-cpdef cnp.ndarray[cnp.double_t, ndim=2] mst_from_data_matrix(
-    cnp.ndarray[cnp.double_t, ndim=2, mode='c'] raw_data,
-    cnp.ndarray[cnp.double_t, ndim=1, mode='c'] core_distances,
+cpdef cnp.ndarray[MST_edge_t, ndim=1, mode='c'] mst_from_data_matrix(
+    const cnp.float64_t[:, ::1] raw_data,
+    const cnp.float64_t[::1] core_distances,
     DistanceMetric dist_metric,
-    cnp.double_t alpha=1.0
 ):
```

Comment:
This function implementation looks highly similar to the oen of `sklearn.cluster.mst_linkage_core`:

https://github.com/scikit-learn/scikit-learn/blob/d8fa96c29828e3ca79ddd5d7466521ac4d95213c/sklearn/cluster/_hierarchical_fast.pyx#L437-L459

Is it possible to factor them in this PR of in another one?

Also, it seems both might be backed by a dedicated `PairwiseDistancesReduction` back-end.

Reply:
They do look similar, but I think the way this function uses `core_distances` makes it a bit different from `mst_linkage_core`.

In any case, if this is a possible refactor, I would prefer to do it in a follow up PR.

Reply:
The overall algorithm's shape is the same, but the specifics are different enough where I don't think we can recycle code directly. Although someone that takes a much closer look may prove me wrong, but as @thomasjpfan said, it would probably be best in a separate PR :)